build.gradle

```diff
@@ -102,28 +93,80 @@ wrapper {
 }
 
 // BuildDate function adapted from http://stackoverflow.com/questions/19172565/how-append-date-build-to-versionnamesuffix-on-gradle
-def buildDate() {
+def buildDate = {
     def df = new SimpleDateFormat('yyyy-MM-dd')
+
     df.setTimeZone(TimeZone.getTimeZone('UTC'))
+
     return df.format(new Date())
 }
 
 gitProperties {
-    customProperty 'buildDate', "${buildDate()}"
+    customProperty 'buildDate', buildDate
     customProperty 'develop', 'true'
 }
 
-task buildLinuxPackage(type: Exec) {
+def stringify(string) {
+    return '"' + string + '"';
+}
+
+def appVersion = {
+    return version
+            .replaceAll("\\-", ".")
+            .replaceAll("alpha", "0")
+            .replaceAll("beta", "0")
+            .replaceAll("release", "0")
+}
+
+task buildDebPackage(type: Exec) {
     commandLine 'jpackage',
             '--verbose',
-            '--name', 'Rails (18xx)',
-            '--description', '"Rails is a moderator program for 18xx games"',
+            '--type', 'deb',
+            '--name', stringify(appName),
```

Comment:
The groovy/gradle way would be to specify this a GString, hence "${appName}"

Reply:
This automatically adds `"` around the string?

Reply:
Not sure it would be any different than others like your main-jar declaration a couple of lines above?

Reply:
The issue with these strings is that they can contain spaces. Because we are using the strings as arguments in a command line we need to wrap them in `"` to tell the shell that it is a single argument and not a list of arguments.

Reply:
Yes, understand the reasoning, just don't know how `commandLine` is dealing with arguments as I've never needed to call out to an external command like this.

FYI, I suspect this will fail on the Mac as jpackage is not exported to /usr/bin/ like java is. jpackage only exists in $JAVA_HOME/bin/jpackage

Reply:
It seems like it is working. I think the reason is that `jpackage` is registered in the system `PATH` variable

Reply:
`> A problem occurred starting process 'command 'jpackage''`

This is on a Mac, using `gradle buildDmgPackage` using JDK 14

Reply:
I think this is a configuration issue on your side. On the Mac VM using GitHub Actions it is working, the folder contains the following files after calling `./gradlew buildDmgPackage`:

```
total 473408
-rw-r--r--   1 runner  staff    26K Apr 10 22:11 1835_20200301_2250_Hakko.rails
-rw-r--r--   1 runner  staff   8.4K Apr 10 22:11 1837_20181226_1653_a.rails
-rw-r--r--   1 runner  staff    13K Apr 10 22:11 1837_20181226_1657_a.rails
-rw-r--r--   1 runner  staff   136M Apr 10 22:15 18xx.log
-rwxr-xr-x   1 runner  staff    18K Apr 10 22:11 LICENSE
-rwxr-xr-x   1 runner  staff   952B Apr 10 22:11 README.md
drwxr-xr-x  10 runner  staff   320B Apr 10 22:14 build
-rw-r--r--   1 runner  staff   5.3K Apr 10 22:11 build.gradle
drwxr-xr-x   3 runner  staff    96B Apr 10 22:11 gradle
-rw-r--r--   1 runner  staff     0B Apr 10 22:11 gradle.properties
-rw-r--r--   1 runner  staff   932B Apr 10 22:11 gradle.yml
-rwxr-xr-x   1 runner  staff   5.6K Apr 10 22:11 gradlew
-rw-r--r--   1 runner  staff   2.9K Apr 10 22:11 gradlew.bat
drwxr-xr-x   7 runner  staff   224B Apr 10 22:11 native
-rw-r--r--   1 runner  staff    16K Apr 10 22:11 pom.xml
-rw-r--r--@  1 runner  staff    87M Apr 10 22:18 rails-18xx-2.2.0.dmg
-rw-r--r--   1 runner  staff    26B Apr 10 22:11 settings.gradle
drwxr-xr-x   6 runner  staff   192B Apr 10 22:11 src
```

You can see that a `rails-18xx-2.2.0.dmg` file has been created. I think most likely `jpackage` is not added to your system path.

Reply:
I'm pretty sure this is not an issue specific to my Mac-- JDK installs on Macs are consistent in location and jpackage is *not* part of the default path. I can't speak to the Mac VM running under GitHub actions but I suspect that that may be a custom JDK install.

Reply:
How have you installed JDK 14? Have you done any additional configuration changes?
I know that switching between different JDK versions can be an issue on both Windows and Linux because you need to watch out that all paths (`java`, `javac`, `JAVA_HOME`, etc.) are changed correctly. What output do you get if you fetch the version of a more obscure application bundled with the JDK (for example `jjs -version`)?

Independent I don't think I can help you here. The issue should be fixed if you add `jpackage` to your system or user path, then it should be detectable when you run `jpackage` in your terminal.

Reply:
Yes. No. Switching versions on a Mac is simple and yes, I'm currently using JDK 14 for this test.

Applications that are bundled with JRE/JDKs on a Mac have it built into the application-- the JRE/JDK is not exposed.

Adding jpackage to a user's PATH is not an appropriate solution-- the user should be able to check out the code and build via ./gradlew without needing to do anything special, other than obviously making sure an appropriate version of the JDK is installed-- which on a Mac is managed by the packages released by Oracle or OpenJDK or Adopt and by default is exposed on the command line without any additional effort by the user.

The plugin that I was playing with was able to find jpackage without an issue so obviously it's right where it's expected to be even if it's not on my PATH :) This is the problem we have with switching to calling out to command line tools directly :( As mentioned earlier, we should really be making the call out relative to JAVA_HOME as that then removes any dependencies on a user's PATH AND given that JAVA_HOME is a well-known env var shouldn't run into any issues as a user that doesn't have JAVA_HOME set is going to have much bigger problems doing anything with Java.

Reply:
I had my own experiences with `JAVA_HOME` set wrong or not set at all. I wouldn't expect this to be set correctly on all systems. Also we are talking about a developer setup here. If we add to our documentation that `jpackage` needs to be set to build the installers locally I think this is fine. The normal end-user normally only downloads an installer from the vendor of the software. In our case the user should download the installer from https://github.com/Rails-18xx/Rails/releases

Reply:
Independent of the discussion:

This is currently only the first version of the `jpackage` integration. While it may not be perfect it does its job, namely it allows the creation of a native installer for all major platforms.

In the future we can extend this in multiple directions if we want to:
-  change to the `jpackage` gradle plugin if we can get it working
-  add some extra checks to our current tasks. Possible checks are for example:
   -  checking for a compatible operating system (e.g. `deb` can only be built on a linux system)
   -  checking/searching for a `jpackage` installation

In my opinion we don't need to fix this issue in this PR.

Also, please be aware that I can't test the scripts for Windows and MacOS outside `GitHub Actions`, where it is apparently working.

Reply:
On Macs the JDK version is managed by the OS, including setting JAVA_HOME. So yes, I would expect this to be set correctly on all Macs. And yes, we are talking about a developer setup here. 

I still disagree, we should fix the build so that a user does not need to do anything out of the ordinary. Needing a user to do something other than make sure JAVA_HOME is set correctly is out-of-the-norm with development projects in this day and age.

Reply:
Let us take two steps back, I don't think I can follow you.

How are you managing your JDK installations? According to https://stackoverflow.com/questions/52524112/how-do-i-install-java-on-mac-osx-allowing-version-switching there are a lot of options to manage your JDK installations. Depending on your choices I guess you can face different difficulties.

Which applications are added to your path when you switch your JDK? I expect at least `java` and `javac` are added, otherwise quite a lot of things are difficult to do.

What exactly do you expect me to do? I don't own a mac. This means that every change I do results in me stumbling in the dark. It makes more sense if you add the necessary parts to the `dmg` creation task, compared to if I try to add these parts.

Reply:
The "official" JDK path-- either with Oracle's package or OpenJDK (or Adopt). Any other solution as listed on that page such as brew, MacPorts, etc would be out of scope of our support.

My suggestion is to make ALL the jpackage executions work based off of JAVA_HOME. If you get that working locally for your env in Linux, given that JAVA_HOME is pretty much a requirement for any dev env, then it should work fine for Mac, for Windows. Nothing super complex or tricky.

Reply:
@jklap can you try with the current version again please? It makes use of `JAVA_HOME` to try to detect `jpackage`